ona/authentication.py

```diff
@@ -0,0 +1,87 @@
+import jwt
+from django.conf import settings
+from django.contrib.auth.models import User
+from django.core.signing import BadSignature
+from django.utils.translation import ugettext as _
+from django.utils import timezone
+from django.shortcuts import get_object_or_404
+from rest_framework import exceptions
+from rest_framework.authentication import (
+    TokenAuthentication, get_authorization_header
+)
+from rest_framework.authtoken.models import Token
+
+
+JWT_SECRET_KEY = getattr(settings, 'JWT_SECRET_KEY', 'jwt')
+JWT_ALGORITHM = getattr(settings, 'JWT_ALGORITHM', 'HS256')
+
+
+def encode_payload(payload):
+    encoded_payload = jwt.encode(
+        payload, JWT_SECRET_KEY, algorithm=JWT_ALGORITHM
+    )
+
+    return encoded_payload
+
+
+def decode_payload(payload):
+    decoded_payload = jwt.decode(
+        payload, JWT_SECRET_KEY, algorithms=[JWT_ALGORITHM]
+    )
+
+    return decoded_payload
+
+
+def get_api_token(json_web_token):
+    """Get API Token from JSON Web Token"""
+    # having the JWT variables here allows the values to be mocked easily as
+    # oppossed to being on the global scope. At the moment they are set
+    # globally mainly because there isn't a test for it
+    try:
+        jwt_payload = decode_payload(json_web_token)
+        api_token = get_object_or_404(Token, key=jwt_payload.get('api-token'))
+
+        return api_token
+    except BadSignature as e:
+        raise exceptions.AuthenticationFailed(_(u'Bad Signature: %s' % e))
+    except jwt.DecodeError as e:
+        raise exceptions.AuthenticationFailed(_(u'JWT DecodeError: %s' % e))
+
+
+class JWTAuthentication(TokenAuthentication):
+    model = Token
+
+    def authenticate(self, request):
+        try:
+            cookie_jwt = request.get_signed_cookie(
+                '__enketo', salt=getattr(settings, 'ENKETO_API_SALT')
+            )
+            api_token = get_api_token(cookie_jwt)
+
+            if getattr(api_token, 'user'):
+                return api_token.user, api_token
+        except self.model.DoesNotExist:
+            raise exceptions.AuthenticationFailed(_(u'Invalid token'))
+        except KeyError:
+            pass
+        except BadSignature:
+            # if the cookie wasn't signed it means zebra might have
+            # generated it
+            cookie_jwt = request.COOKIES.get('__enketo')
```

Comment:
I get your concern. The main reason I used the enketo cookie is I was trying to avoid creating a new cookie which has exactly the same value as the `__enketo` cookie but I guess it won't hurt creating that new cookie.